In [1]:
import pandas as pd 
import numpy as np 
import seaborn as sns

df = sns.load_dataset('tips')

In [7]:
df.groupby('day')['tip'].mean()

/var/folders/vr/dr1s8fb178gfh3rh2tzhxjth0000gp/T/ipykernel_92261/677977839.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby('day')['tip']


In [4]:
df.groupby('day')['tip'].agg(['count', 'mean', 'median', 'std'])

/var/folders/vr/dr1s8fb178gfh3rh2tzhxjth0000gp/T/ipykernel_92261/3695051188.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby('day')['tip'].agg(['count', 'mean', 'median', 'std'])


,count,mean,median,std
day,,,,
Thur,62,2.771452,2.305,1.240223
Fri,19,2.734737,3.000,1.019577
Sat,87,2.993103,2.750,1.631014
Sun,76,3.255132,3.150,1.234880


In [8]:
df.groupby(['day', 'time'])['tip'].mean()

/var/folders/vr/dr1s8fb178gfh3rh2tzhxjth0000gp/T/ipykernel_92261/3498249342.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby(['day', 'time'])['tip'].mean()


day   time  
Thur  Lunch     2.767705
      Dinner    3.000000
Fri   Lunch     2.382857
      Dinner    2.940000
Sat   Lunch          NaN
      Dinner    2.993103
Sun   Lunch          NaN
      Dinner    3.255132
Name: tip, dtype: float64

In [10]:
df.groupby('day').agg({
    'tip': 'mean',
    'total_bill': 'sum',
    'size': 'count'
})

/var/folders/vr/dr1s8fb178gfh3rh2tzhxjth0000gp/T/ipykernel_92261/2431572831.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby('day').agg({


,tip,total_bill,size
day,,,
Thur,2.771452,1096.33,62
Fri,2.734737,325.88,19
Sat,2.993103,1778.40,87
Sun,3.255132,1627.16,76


In [28]:
df.groupby('day')['tip'].apply(lambda x: x.max() - x.min()).reset_index()

/var/folders/vr/dr1s8fb178gfh3rh2tzhxjth0000gp/T/ipykernel_92261/3256190220.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby('day')['tip'].apply(lambda x: x.max() - x.min()).reset_index()


,day,tip
0,Thur,5.45
1,Fri,3.73
2,Sat,9.00
3,Sun,5.49


In [12]:
df.groupby('day').filter(lambda g: len(g) > 50)

/var/folders/vr/dr1s8fb178gfh3rh2tzhxjth0000gp/T/ipykernel_92261/3823121822.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby('day').filter(lambda g: len(g) > 50)


,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,Female,No,Sun,Dinner,2
1,10.34,1.66,Male,No,Sun,Dinner,3
2,21.01,3.50,Male,No,Sun,Dinner,3
3,23.68,3.31,Male,No,Sun,Dinner,2
4,24.59,3.61,Female,No,Sun,Dinner,4
...,...,...,...,...,...,...,...
239,29.03,5.92,Male,No,Sat,Dinner,3
240,27.18,2.00,Female,Yes,Sat,Dinner,2
241,22.67,2.00,Male,Yes,Sat,Dinner,2
242,17.82,1.75,Male,No,Sat,Dinner,2


<bound method IndexOpsMixin.value_counts of 0       Sun
1       Sun
2       Sun
3       Sun
4       Sun
       ... 
239     Sat
240     Sat
241     Sat
242     Sat
243    Thur
Name: day, Length: 244, dtype: category
Categories (4, object): ['Thur', 'Fri', 'Sat', 'Sun']>

In [17]:
df.groupby('day').agg(
    avg_tip=('tip', 'mean'),
    total_revenue=('total_bill', 'sum'),
    visits=('size', 'count')
)

/var/folders/vr/dr1s8fb178gfh3rh2tzhxjth0000gp/T/ipykernel_92261/2295858802.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby('day').agg(


,avg_tip,total_revenue,visits
day,,,
Thur,2.771452,1096.33,62
Fri,2.734737,325.88,19
Sat,2.993103,1778.40,87
Sun,3.255132,1627.16,76


In [20]:
df['day_avg_tip'] = df.groupby('day')['tip'].transform('mean').round(2)

/var/folders/vr/dr1s8fb178gfh3rh2tzhxjth0000gp/T/ipykernel_92261/3106703205.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df['day_avg_tip'] = df.groupby('day')['tip'].transform('mean').round(2)


In [21]:
df

,total_bill,tip,sex,smoker,day,time,size,day_avg_tip
0,16.99,1.01,Female,No,Sun,Dinner,2,3.26
1,10.34,1.66,Male,No,Sun,Dinner,3,3.26
2,21.01,3.50,Male,No,Sun,Dinner,3,3.26
3,23.68,3.31,Male,No,Sun,Dinner,2,3.26
4,24.59,3.61,Female,No,Sun,Dinner,4,3.26
...,...,...,...,...,...,...,...,...
239,29.03,5.92,Male,No,Sat,Dinner,3,2.99
240,27.18,2.00,Female,Yes,Sat,Dinner,2,2.99
241,22.67,2.00,Male,Yes,Sat,Dinner,2,2.99
242,17.82,1.75,Male,No,Sat,Dinner,2,2.99


In [26]:
result = df.groupby('day')['tip'].mean()
result.reset_index()

/var/folders/vr/dr1s8fb178gfh3rh2tzhxjth0000gp/T/ipykernel_92261/3394118223.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  result = df.groupby('day')['tip'].mean()


,day,tip
0,Thur,2.771452
1,Fri,2.734737
2,Sat,2.993103
3,Sun,3.255132


In [42]:
transactions = pd.DataFrame({
    'client_id': [1, 2, 2, 3, 5],
    'amount': [100, 200, 150, 300, 400]
})

customers = pd.DataFrame({
    'customer_id': [1, 2, 3, 4],
    'name': ['Alice', 'Bob', 'Carol', 'Dave']
})

In [33]:
pd.merge(customers, transcations, on='customer_id')

,customer_id,name,amount
0,1,Alice,100
1,2,Bob,200
2,2,Bob,150
3,3,Carol,300


In [38]:
pd.merge(customers, transactions, on='customer_id', how='left')

,customer_id,name,amount
0,1,Alice,100.0
1,2,Bob,200.0
2,2,Bob,150.0
3,3,Carol,300.0
4,4,Dave,NaN


In [40]:
pd.merge(customers, transactions, on='customer_id', how='outer')

,customer_id,name,amount
0,1,Alice,100.0
1,2,Bob,200.0
2,2,Bob,150.0
3,3,Carol,300.0
4,4,Dave,NaN
5,5,NaN,400.0


In [43]:
pd.merge(customers, transactions, left_on='customer_id', right_on='client_id')

,customer_id,name,client_id,amount
0,1,Alice,1,100
1,2,Bob,2,200
2,2,Bob,2,150
3,3,Carol,3,300


In [47]:
pivot = df.pivot_table(
    index='day',
    columns='time',
    values='tip',
    aggfunc='mean',
    fill_value=69
).round(2)
pivot

/var/folders/vr/dr1s8fb178gfh3rh2tzhxjth0000gp/T/ipykernel_92261/771276215.py:1: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  pivot = df.pivot_table(


time,Lunch,Dinner
day,,
Thur,2.77,3.00
Fri,2.38,2.94
Sat,69.00,2.99
Sun,69.00,3.26


In [52]:
df.pivot_table(
    index='day',
    columns='time',
    values=['tip', 'total_bill'],
    aggfunc='mean'
).round(2)

/var/folders/vr/dr1s8fb178gfh3rh2tzhxjth0000gp/T/ipykernel_92261/168233999.py:1: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  df.pivot_table(


tip        total_bill       
time Lunch Dinner      Lunch Dinner
day                                
Thur  2.77   3.00      17.66  18.78
Fri   2.38   2.94      12.85  19.66
Sat    NaN   2.99        NaN  20.44
Sun    NaN   3.26        NaN  21.41

In [56]:
pd.crosstab(df['day'], df['time'])

time,Lunch,Dinner
day,,
Thur,61,1
Fri,7,12
Sat,0,87
Sun,0,76


In [59]:
pd.crosstab(df['day'], df['time'], normalize='index')

time,Lunch,Dinner
day,,
Thur,0.983871,0.016129
Fri,0.368421,0.631579
Sat,0.000000,1.000000
Sun,0.000000,1.000000


In [60]:
pd.concat([df1, df2], axis=0)

NameError: name 'df1' is not defined

In [61]:
import pandas as pd 
import seaborn as sns 

df = sns.load_dataset('tips')


In [8]:
df['tip_percent'] = df['tip'] / df['total_bill'] * 100
df

,total_bill,tip,sex,smoker,day,time,size,tip_percent
0,16.99,1.01,Female,No,Sun,Dinner,2,5.944673
1,10.34,1.66,Male,No,Sun,Dinner,3,16.054159
2,21.01,3.50,Male,No,Sun,Dinner,3,16.658734
3,23.68,3.31,Male,No,Sun,Dinner,2,13.978041
4,24.59,3.61,Female,No,Sun,Dinner,4,14.680765
...,...,...,...,...,...,...,...,...
239,29.03,5.92,Male,No,Sat,Dinner,3,20.392697
240,27.18,2.00,Female,Yes,Sat,Dinner,2,7.358352
241,22.67,2.00,Male,Yes,Sat,Dinner,2,8.822232
242,17.82,1.75,Male,No,Sat,Dinner,2,9.820426


In [78]:
df.groupby('day')['tip_percent'].mean().round(2).reset_index()

/var/folders/vr/dr1s8fb178gfh3rh2tzhxjth0000gp/T/ipykernel_92261/463677313.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby('day')['tip_percent'].mean().round(2).reset_index()


,day,tip_percent
0,Thur,16.13
1,Fri,16.99
2,Sat,15.31
3,Sun,16.69


In [80]:
# 1. Какой средний tip_percent по дням недели?
#    Average tip percentage per day

# 2. Какой ДЕНЬ + ВРЕМЯ даёт самые высокие чаевые в процентах?
#    Highest average tip_percent for (day × time) combination.
#    Используй groupby с двумя колонками

# 3. Посчитай одной командой по дням: count, mean, median, std чаевых
#    Multiple aggregations at once

# 4. Для каждого дня: общий оборот (sum total_bill), средний чек и количество визитов.
#    Используй named aggregation (самый красивый синтаксис):
#    df.groupby('day').agg(
#        revenue=('total_bill', 'sum'),
#        avg_bill=('total_bill', 'mean'),
#        visits=('size', 'count')
#    )

# 5. Создай новую колонку `day_avg_tip` — средний tip_percent в день визита.
#    Для каждой строки — своё значение.
#    Подсказка: groupby + transform

In [94]:
(df.groupby(['day', 'time'])['tip_percent']
    .mean()
    .round(2)
    .sort_values(ascending=False)
    .reset_index())

/var/folders/vr/dr1s8fb178gfh3rh2tzhxjth0000gp/T/ipykernel_92261/4093468427.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  (df.groupby(['day', 'time'])['tip_percent']


,day,time,tip_percent
0,Fri,Lunch,18.88
1,Sun,Dinner,16.69
2,Thur,Lunch,16.13
3,Thur,Dinner,15.97
4,Fri,Dinner,15.89
5,Sat,Dinner,15.31
6,Sat,Lunch,NaN
7,Sun,Lunch,NaN


In [95]:
(df.groupby(['day', 'time'])['tip_percent']
    .mean()
    .round(2)
    .idxmax())

/var/folders/vr/dr1s8fb178gfh3rh2tzhxjth0000gp/T/ipykernel_92261/1238622254.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  (df.groupby(['day', 'time'])['tip_percent']


('Fri', 'Lunch')

the highest average tip percent with combination day x time is fridayxlunch iwth 18.88% percent, which higher for 2% then the second highest one


In [101]:
df.groupby('day')['tip'].agg(['count', 'mean', 'median', 'std']).round(2)

/var/folders/vr/dr1s8fb178gfh3rh2tzhxjth0000gp/T/ipykernel_92261/3773285886.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby('day')['tip'].agg(['count', 'mean', 'median', 'std']).round(2)


,count,mean,median,std
day,,,,
Thur,62,2.77,2.30,1.24
Fri,19,2.73,3.00,1.02
Sat,87,2.99,2.75,1.63
Sun,76,3.26,3.15,1.23


In [1]:
# 4. Для каждого дня: общий оборот (sum total_bill), средний чек и количество визитов.
#    Используй named aggregation (самый красивый синтаксис):
#    df.groupby('day').agg(
#        revenue=('total_bill', 'sum'),
#        avg_bill=('total_bill', 'mean'),
#        visits=('size', 'count')
#    )

# 5. Создай новую колонку `day_avg_tip` — средний tip_percent в день визита.
#    Для каждой строки — своё значение.
#    Подсказка: groupby + transform

In [7]:
df.groupby('day').agg(
    revenue=('total_bill', 'sum'),
    abg_bill=('total_bill', 'mean'),
    visits=('size', 'count')
).round(2)

df

/var/folders/vr/dr1s8fb178gfh3rh2tzhxjth0000gp/T/ipykernel_17234/1388468942.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby('day').agg(


,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,Female,No,Sun,Dinner,2
1,10.34,1.66,Male,No,Sun,Dinner,3
2,21.01,3.50,Male,No,Sun,Dinner,3
3,23.68,3.31,Male,No,Sun,Dinner,2
4,24.59,3.61,Female,No,Sun,Dinner,4
...,...,...,...,...,...,...,...
239,29.03,5.92,Male,No,Sat,Dinner,3
240,27.18,2.00,Female,Yes,Sat,Dinner,2
241,22.67,2.00,Male,Yes,Sat,Dinner,2
242,17.82,1.75,Male,No,Sat,Dinner,2


In [28]:
df['day_avg_tip'] = df.groupby('day')['tip_percent'].transform('mean').round(2)


/var/folders/vr/dr1s8fb178gfh3rh2tzhxjth0000gp/T/ipykernel_17234/472946420.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df['day_avg_tip'] = df.groupby('day')['tip_percent'].transform('mean').round(2)


In [29]:
df

,total_bill,tip,sex,smoker,day,time,size,tip_percent,day_avg_tip
0,16.99,1.01,Female,No,Sun,Dinner,2,5.944673,16.69
1,10.34,1.66,Male,No,Sun,Dinner,3,16.054159,16.69
2,21.01,3.50,Male,No,Sun,Dinner,3,16.658734,16.69
3,23.68,3.31,Male,No,Sun,Dinner,2,13.978041,16.69
4,24.59,3.61,Female,No,Sun,Dinner,4,14.680765,16.69
...,...,...,...,...,...,...,...,...,...
239,29.03,5.92,Male,No,Sat,Dinner,3,20.392697,15.32
240,27.18,2.00,Female,Yes,Sat,Dinner,2,7.358352,15.32
241,22.67,2.00,Male,Yes,Sat,Dinner,2,8.822232,15.32
242,17.82,1.75,Male,No,Sat,Dinner,2,9.820426,15.32


In [37]:
# 6. Создай искусственную вторую таблицу:
day_info = pd.DataFrame({
    'day': ['Thur', 'Fri', 'Sat', 'Sun'],
    'is_weekend': [False, False, True, True],
    'day_rank': [4, 5, 6, 7]
})
day_info
# Объедини df с day_info через merge по колонке 'day'. Используй LEFT join.
# Merge df with day_info using LEFT join on 'day'.

pd.merge(df, day_info, on='day', how='left')
# Проверь: df.shape до и после — должно остаться столько же строк


# 7. Посчитай средний tip_percent для выходных vs будней
# Используй новую колонку is_weekend из задачи 6


,total_bill,tip,sex,smoker,day,time,size,tip_percent,day_avg_tip,is_weekend,day_rank
0,16.99,1.01,Female,No,Sun,Dinner,2,5.944673,16.69,True,7
1,10.34,1.66,Male,No,Sun,Dinner,3,16.054159,16.69,True,7
2,21.01,3.50,Male,No,Sun,Dinner,3,16.658734,16.69,True,7
3,23.68,3.31,Male,No,Sun,Dinner,2,13.978041,16.69,True,7
4,24.59,3.61,Female,No,Sun,Dinner,4,14.680765,16.69,True,7
...,...,...,...,...,...,...,...,...,...,...,...
239,29.03,5.92,Male,No,Sat,Dinner,3,20.392697,15.32,True,6
240,27.18,2.00,Female,Yes,Sat,Dinner,2,7.358352,15.32,True,6
241,22.67,2.00,Male,Yes,Sat,Dinner,2,8.822232,15.32,True,6
242,17.82,1.75,Male,No,Sat,Dinner,2,9.820426,15.32,True,6


In [44]:
df = pd.merge(df, day_info, on='day', how='left')

df.groupby('is_weekend').agg(
    avg_tip_percent=('tip_percent', 'mean'),
    visits=('tip_percent', 'count')
).round(2)


,avg_tip_percent,visits
is_weekend,,
False,16.33,81
True,15.96,163


In [60]:
df.pivot_table(
    index='day',
    columns='time',
    values='tip_percent',
    aggfunc='mean',
    fill_value=0
).round(2)

/var/folders/vr/dr1s8fb178gfh3rh2tzhxjth0000gp/T/ipykernel_17234/1795183722.py:1: FutureWarning: The default value of observed=False is deprecated and will change to observed=True in a future version of pandas. Specify observed=False to silence this warning and retain the current behavior
  df.pivot_table(


time,Lunch,Dinner
day,,
Fri,18.88,15.89
Sat,0.00,15.32
Sun,0.00,16.69
Thur,16.13,15.97


In [55]:
pd.crosstab(df['sex'], df['day'])

day,Fri,Sat,Sun,Thur
sex,,,,
Male,10,59,58,30
Female,9,28,18,32


In [59]:
pd.crosstab(df['sex'], df['smoker'], normalize='index')

smoker,Yes,No
sex,,
Male,0.382166,0.617834
Female,0.379310,0.620690


In [71]:
df.groupby(['day', 'time'])['tip_percent'].mean().round(2).sort_values(ascending=False).idxmax()




/var/folders/vr/dr1s8fb178gfh3rh2tzhxjth0000gp/T/ipykernel_17234/308312345.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby(['day', 'time'])['tip_percent'].mean().round(2).sort_values(ascending=False).idxmax()


('Fri', 'Lunch')

In [77]:
df['segment'] = pd.cut(
    df['tip_percent'],
    bins=[0, 15, 20, 100],
    labels=['low', 'medium', 'high']
)

In [98]:
df['segment'].value_counts(normalize=True).round(3) * 100

segment
low       44.7
medium    39.3
high      16.0
Name: proportion, dtype: float64

In [101]:
df['segment'].value_counts() / len(df) * 100

segment
low       44.672131
medium    39.344262
high      15.983607
Name: count, dtype: float64

In [90]:
high = df[df['segment'] == 'high']

In [94]:
high['sex'].value_counts(normalize=True) * 100
high['day'].value_counts(normalize=True) * 100
high['time'].value_counts(normalize=True) * 100
high['smoker'].value_counts(normalize=True) * 100

smoker
No     58.974359
Yes    41.025641
Name: proportion, dtype: float64

In [103]:
for col in high[['sex', 'day', 'time', 'smoker']]:
    print(f'\n=== {col} ===')
    print(high[col].value_counts(normalize=True).round(3))


=== sex ===
sex
Male      0.615
Female    0.385
Name: proportion, dtype: float64

=== day ===
day
Sun     0.385
Sat     0.256
Thur    0.256
Fri     0.103
Name: proportion, dtype: float64

=== time ===
time
Dinner    0.692
Lunch     0.308
Name: proportion, dtype: float64

=== smoker ===
smoker
No     0.59
Yes    0.41
Name: proportion, dtype: float64


In [110]:
df['is_vip'] = (df['segment'] == 'VIP').astype(int)

In [111]:
df['is_vip'].mean()

np.float64(0.0)

# Conclusions

